### PR TITLE
Reorganizing and adding helpers.

### PR DESCRIPTION
**Pull request by drystone**

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -691,13 +691,13 @@ impl Gdbm {
...
+        let mut elem_ofs = (elem_ofs + 1) % bucket_elems;
         while elem_ofs != last_ofs && bucket.tab[elem_ofs].hash != 0xffffffff {
             let home = (bucket.tab[elem_ofs].hash as usize) % bucket_elems;
             if (last_ofs < elem_ofs && (home <= last_ofs || home > elem_ofs))
                 || (last_ofs > elem_ofs && home <= last_ofs && home > elem_ofs)
             {
-                bucket.tab[last_ofs] = bucket.tab[elem_ofs].clone();
-                bucket.tab[elem_ofs].hash = 0xffffffff;
+                bucket.tab.swap(elem_ofs, last_ofs);
                 last_ofs = elem_ofs;
```
[the remark this answers is not in the document]
Comment (drystone) [reply src/lib.rs:650]: IIRC this line was missing:
```
  elem_loc = (elem_loc + 1) % dbf->header->bucket_elems;
```
so the loop never got started.

File: src/lib.rs [hunk 2]
```diff
@@ -466,38 +465,25 @@ impl Gdbm {
...
+    // virtually allocate N blocks of data, at end of db file
+    fn extend(&mut self, size: u32) -> io::Result<(u64, u32)> {
+        let offset = self.header.next_block;
+        let length = match size % self.header.block_sz {
+            0 => size / self.header.block_sz,
+            _ => size / self.header.block_sz + 1,
+        } * self.header.block_sz;
 
+        self.header.next_block += length as u64;
+        self.f.set_len(self.header.next_block)?;
```
[the remark this answers is not in the document]
Comment (drystone) [reply src/lib.rs:450]: I'm thinking that a check here to make sure there's enough disk space is worth the effort. It only happens when new blocks are added to the end of the file, and out of space is the most likely error we'll come across.

File: src/lib.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (jgarzik) [reply src/lib.rs:646]: OK.  To resolve without adding additional regressions, it sounds like the path forward is to NOT merge this, and put in a bug fix-only PR and commit?

**Comment by jgarzik** [comment]
Contra the two decisions just posted, I want to compliment you on breaking up your changes into logical commits.  It is very refreshing and greatly aids in both `git bisect`, testing and review.

File: src/lib.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (drystone) [reply src/lib.rs:646]: This is an individual bugfix commit.

https://github.com/jgarzik/gdbm-native-rs/pull/68/commits/7b27faa20a5144f04c3332bc1ecfec3b5504bbda

I can put it in a separate PR if necessary, but it won't make any difference to the commit history.

**Comment by drystone** [comment]
Thanks @jgarzik :)

File: src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (drystone) [reply src/lib.rs:442]: Sure, I'll revert it.

File: src/lib.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (jgarzik) [reply src/lib.rs:646]: That bugfix commit fails to set hash_value to -1, an unwanted regression, which generated my request for an updated commit or PR.

[the remark this answers is not in the document]
Comment (drystone) [reply src/lib.rs]: No - that's not a bug.

Conceptually we are moving the empty element along the array.

That line simply swaps the 2 array elements. `last_ofs` is the "empty" element with -1 hash. So the element at `elem_ofs` is implicitly made empty (the empty element is moved there).

**Pull request merged**